flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/strategy/LazyFromSourcesSchedulingStrategyTest.java

```diff
@@ -0,0 +1,342 @@
+
+		ResultPartitionType[] partitionTypes = new ResultPartitionType[jobVertexCnt - 1];
+		InputDependencyConstraint[] inputDependencyConstraints = new InputDependencyConstraint[jobVertexCnt];
+		partitionTypes[0] = BLOCKING;
+		inputDependencyConstraints[0] = ANY;
+		inputDependencyConstraints[1] = ANY;
+
+		initVerticesAndPartitions(schedulingTopology, partitionTypes, inputDependencyConstraints, vertices, partitions);
+
+		for (int i = 0; i < taskCnt; i++) {
+			connectVerticesToPartition(vertices[0][i], vertices[1][i], partitions[0][i]);
+		}
+
+		TestingSchedulerOperations testingSchedulerOperation = new TestingSchedulerOperations();
+		LazyFromSourcesSchedulingStrategy schedulingStrategy = new LazyFromSourcesSchedulingStrategy(
+			testingSchedulerOperation,
+			schedulingTopology);
+
+		schedulingStrategy.startScheduling();
+
+		schedulingStrategy.onExecutionStateChange(vertices[0][0].getId(), ExecutionState.FINISHED);
+		schedulingStrategy.onExecutionStateChange(vertices[0][1].getId(), ExecutionState.FINISHED);
+		schedulingStrategy.onExecutionStateChange(vertices[0][2].getId(), ExecutionState.FINISHED);
+		Collection<ExecutionVertexDeploymentOption> scheduledVertices = testingSchedulerOperation.getScheduledVertices().get(1);
+		Set<ExecutionVertexID> toBeScheduledVertices = Arrays.stream(vertices[1])
+			.map(TestingSchedulingExecutionVertex::getId).collect(Collectors.toSet());
+		assertThat(getExecutionVertexIdsFromDeployOptions(scheduledVertices),
+			containsInAnyOrder(toBeScheduledVertices.toArray()));
+	}
+
+	/**
+	 * Tests that when all the input partitions are ready will start available downstream {@link BLOCKING} vertices.
+	 * vertex#0    vertex#1
+	 *       \     /
+	 *        \   /
+	 *         \ /
+	 *  (BLOCKING, ALL)
+	 *     vertex#2
+	 */
+	@Test
+	public void testBLOCKINGALLExecutionStateChange() {
+		final int jobVertexCnt = 3;
+		final int taskCnt = 2;
+		TestingSchedulingTopology schedulingTopology = new TestingSchedulingTopology();
+		TestingSchedulingExecutionVertex[][] vertices = new TestingSchedulingExecutionVertex[jobVertexCnt][taskCnt];
+		TestingSchedulingResultPartition[][] partitions = new TestingSchedulingResultPartition[jobVertexCnt - 1][taskCnt];
+
+		ResultPartitionType[] partitionTypes = new ResultPartitionType[jobVertexCnt - 1];
+		InputDependencyConstraint[] inputDependencyConstraints = new InputDependencyConstraint[jobVertexCnt];
+		partitionTypes[0] = BLOCKING;
+		partitionTypes[1] = BLOCKING;
+		inputDependencyConstraints[0] = ANY;
+		inputDependencyConstraints[1] = ANY;
+		inputDependencyConstraints[2] = InputDependencyConstraint.ALL;
+
+		initVerticesAndPartitions(schedulingTopology, partitionTypes, inputDependencyConstraints, vertices, partitions);
+
+		for (int i = 0; i < taskCnt; i++) {
+			for (int j = 0; j < taskCnt; j++) {
+				connectVerticesToPartition(vertices[0][i], vertices[2][i], partitions[0][j]);
```

Comment:
It looks like that `vertices[0][0]` and `vertices[0][1]` produce the same partitions. This is not a legal topology. Any intermediate result partition has exactly one producer.

Reply:
OK